pkg/hash/configmap.go

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ */
+
+package hash
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/pkg/errors"
+)
+
+// hash package purpose is to compute a ConfigMap hash
+// that will be attached as an annotation to workload resources (DaemonSet, Deployment, etc.)
+// in order to allow them to track ConfigMap changes
+// more about this technique here: https://blog.questionable.services/article/kubernetes-deployments-configmap-change/
+
+const ConfigMapAnnotation = "configmap.hash"
+
+func ComputeCurrentConfigMap(ctx context.Context, cli client.Client, cm *corev1.ConfigMap) (string, error) {
+	updatedConfigMap := &corev1.ConfigMap{}
+	key := client.ObjectKeyFromObject(cm)
+
+	if err := cli.Get(ctx, key, updatedConfigMap); err != nil {
+		// ConfigMap not created yet, use the data from the manifests
+		if apierrors.IsNotFound(err) {
+			updatedConfigMap = cm
+		} else {
+			return "", errors.Wrapf(err, "could not calculate ConfigMap %q hash", key.String())
+		}
+	}
+	cmHash := ConfigMapData(updatedConfigMap)
+	klog.InfoS("configmap hash calculated", "hash", cmHash)
+	return cmHash, nil
+}
+
+func ConfigMapData(cm *corev1.ConfigMap) string {
+	var dataAsString string
+
+	if cm.Data != nil {
```

Comment:
this is correct, but if `BinaryData` is != nil, this function will silently break. We should at least detect and error out accordingly, because we don't need `BinaryData` now or in the near future, but breaking silently is just too bad

Reply:
> If BinaryData is != nil, this function will silently break.

Sorry for my ignorance, but why that should happen?

Reply:
no, I mean why if in case that  ```BinaryData != nil``` the code will break? I'm not saying that it should never happen that ``BinaryData != nil```, I'm asking why if it's happening it will break the code

Reply:
ah, sorry. Perhaps "break" is too strong of a statement. And also this comment is maybe misplaced, should have been in ComputeCurrentConfigMap

The fact is that if`BinaryData` is not taken into account in the hash, then different ConfigMaps (because BinaryData is different) can have the same hash, which is not what we want.
The problem is the ComputeCurrentConfigMap function whant to compute the hash of a `ConfigMap`, which in general can have a BinaryData, so we either restrict ComputeCurrentConfigMap to consider only the `Data` (maybe renaming it?) or  we need to handle `BinaryData` somehow. Erroring out if `!= nil` is an option, and so is adding the `BinaryData` in the hash. I'm fine with all the approaches, as long as it is handled in `ComputeCurrentConfigMap` - `ConfigMapData` is indeed fine.

Reply:
Now it's clear, thanks!

IMHO ```BinaryData``` is part of data so it should be taken under consideration as well  (ofc when it's not ```nil```) this will make the function more generic and will allow support in other ConfigMap types which do contain ```BinaryData```

Reply:
That's the best solution IMO!